File: uri/strings/palindrome/1941.cpp
#include <iostream>
#include <string>
#include <vector>
#include <limits>
#include <sstream>

using namespace std;

struct Node {
    int special;
    int length;
};

inline bool better(const Node &a, const Node &b) {
    if (a.special != b.special) {
        return a.special > b.special;
    }
    return a.length > b.length;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    string s;
    if (!(cin >> s)) {
        return 0;
    }
    int n = static_cast<int>(s.size());

    vector<bool> special(n, false);
    string rest;
    getline(cin, rest); 
    string line;
    if (!getline(cin, line)) {
        line.clear();
    }
    if (!line.empty()) {
        stringstream ss(line);
        int count = 0;
        ss >> count;
        for (int i = 0; i < count; ++i) {
            int idx;
            ss >> idx;
            if (idx >= 1 && idx <= n) {
                special[idx - 1] = true;
            }
        }
    }

    vector<vector<Node>> dp(n, vector<Node>(n, {0, 0}));

    for (int i = 0; i < n; ++i) {
        dp[i][i] = {special[i] ? 1 : 0, 1};
    }

    for (int len = 2; len <= n; ++len) {
        for (int i = 0; i + len - 1 < n; ++i) {
            int j = i + len - 1;
            Node best = dp[i + 1][j];
            if (better(dp[i][j - 1], best)) {
                best = dp[i][j - 1];
            }
            if (s[i] == s[j]) {
                Node candidate;
                int extra = (special[i] ? 1 : 0) + (special[j] ? 1 : 0);
                if (len == 2) {
                    candidate = {extra, 2};
                } else {
                    candidate.special = dp[i + 1][j - 1].special + extra;
                    candidate.length = dp[i + 1][j - 1].length + 2;
                }
                if (better(candidate, best)) {
                    best = candidate;
                }
            }
            dp[i][j] = best;
        }
    }

    cout << dp[0][n - 1].length << '\n';
    return 0;
}

File: uri/strings/palindrome/1941.rs
use std::io::{self, Read};

#[derive(Clone, Copy)]
struct Node {
    special: i32,
    length: i32,
}

fn better(a: &Node, b: &Node) -> bool {
    if a.special != b.special {
        a.special > b.special
    } else if a.length != b.length {
        a.length > b.length
    } else {
        false
    }
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut lines = input.lines();

    let s_line = match lines.next() {
        Some(line) => line.trim().to_string(),
        None => return,
    };
    let n = s_line.len();
    if n == 0 {
        println!("0");
        return;
    }

    let info_line = lines.next().unwrap_or("").trim().to_string();
    let numbers: Vec<usize> = if info_line.is_empty() {
        Vec::new()
    } else {
        info_line
            .split_whitespace()
            .map(|token| token.parse::<usize>().unwrap())
            .collect()
    };

    let mut special = vec![false; n];
    if !numbers.is_empty() {
        let count = numbers[0];
        for pos in numbers.iter().skip(1).take(count) {
            if *pos >= 1 && *pos <= n {
                special[pos - 1] = true;
            }
        }
    }

    let bytes = s_line.as_bytes();
    let mut dp = vec![vec![Node { special: 0, length: 0 }; n]; n];

    for i in 0..n {
        dp[i][i] = Node {
            special: if special[i] { 1 } else { 0 },
            length: 1,
        };
    }

    for len in 2..=n {
        for i in 0..=n - len {
            let j = i + len - 1;
            let mut best = dp[i + 1][j];
            if better(&dp[i][j - 1], &best) {
                best = dp[i][j - 1];
            }
            if bytes[i] == bytes[j] {
                let extra = (if special[i] { 1 } else { 0 }) + (if special[j] { 1 } else { 0 });
                let candidate = if len == 2 {
                    Node {
                        special: extra,
                        length: 2,
                    }
                } else {
                    Node {
                        special: dp[i + 1][j - 1].special + extra,
                        length: dp[i + 1][j - 1].length + 2,
                    }
                };
                if better(&candidate, &best) {
                    best = candidate;
                }
            }
            dp[i][j] = best;
        }
    }

    println!("{}", dp[0][n - 1].length);
}
